### ISSUE TITLE
Proposal: Allow using ExMachina.Ecto without a repo.

### ISSUE THREAD

**Issue by bamorim** [thoughtbot/ex_machina#355]
Ecto has been splitting the persistence layer a lot lately. As of Ecto 3, `:ecto_sql` is even a different package.

One of the usages of Ecto is just using plain embedded schemas to define some fields. I personally use them a lot in our web layer.
In our most recent project, every mutation method on our controllers are accompanied by an Ecto embedded schema. Our web layer doesn't have any Repository, since it doesn't have any normal schemas, just embedded schemas.

That being said, I don't think `ExMachina.Ecto` should enforce having a `:repo`.

My sugesstion is that if no repo is passed, it defines all methods except the insert ones.
The `params_for` and `string_params_for` are useful even without the Repo (since it strips Ecto related stuff).

So my suggestion is simple (I can open a PR for that easily): just allow repo to not be passed.

It would look something like

```elixir
defmacro __using__(opts) do
  verify_ecto_dep()

  repo = Keyword.get(opts, :repo)

  for_ecto_with_repo = quote do
    use ExMachina.EctoStrategy, repo: unquote(repo)

    def params_with_assocs(factory_name, attrs \\ %{}) do
      ExMachina.Ecto.params_with_assocs(__MODULE__, factory_name, attrs)
    end

    def string_params_with_assocs(factory_name, attrs \\ %{}) do
      ExMachina.Ecto.string_params_with_assocs(__MODULE__, factory_name, attrs)
    end
  end

  for_all = quote do
    use ExMachina

    def params_for(factory_name, attrs \\ %{}) do
      ExMachina.Ecto.params_for(__MODULE__, factory_name, attrs)
    end

    def string_params_for(factory_name, attrs \\ %{}) do
      ExMachina.Ecto.string_params_for(__MODULE__, factory_name, attrs)
    end
  end

  if not is_nil(repo) do
    quote do
      unquote(for_all)
      unquote(for_ecto_with_repo)
    end
  else
    for_all
  end
end
```

What do you guys think?

**Comment by johantell** [comment]
I think using `false` over `nil` could make it more explicit (otherwise I can see cases where one could forgot to add/misspell the `repo` option and get a lot of headaches)

```elixir
use ExMachina.Ecto, repo: false
```

**Comment by germsvel** [comment]
Hi @bamorim, thanks for suggesting this! I do think it makes sense to have `params_for` and `string_params_for` available for people using `ecto` and not `ecto_sql`. I've been trying to think of alternatives to not passing `repo` that would bring more clarity as to what's going on. 

One idea that I had, though I'm not sure if it's better, is to have a separate module called `ExMachina.EctoSchema` (or something like that -- naming is hard). That module could just define similar macros and delegate calls to `ExMachina.Ecto`. So I'm imagining something like this, 

```elixir
defmodule ExMachina.EctoSchema do
  @moduledoc """
  Module for building factories with Ecto

  This module works much like the regular `ExMachina` module, but adds the
  `params_for` and `string_params_for` functions that make working with Ecto
  schemas easier.

  More in-depth examples are in the [README](readme.html).
  """
  defmacro __using__(_) do
    verify_ecto_dep()

    quote do
      use ExMachina

      def params_for(factory_name, attrs \\ %{}) do
        ExMachina.Ecto.params_for(__MODULE__, factory_name, attrs)
      end

      def string_params_for(factory_name, attrs \\ %{}) do
        ExMachina.Ecto.string_params_for(__MODULE__, factory_name, attrs)
      end
    end
  end

  defp verify_ecto_dep do
    unless Code.ensure_loaded?(Ecto) do
      raise "You tried to use ExMachina.EctoSchema, but the Ecto module is not loaded. " <>
              "Please add ecto to your dependencies."
    end
  end
end
```

I don't like that it ties `ExMachina.EctoSchema` to `ExMachina.Ecto` so closely, since changes to `ExMachina.Ecto` now need to keep in mind that `EctoSchema` uses it, but then again, that might not be a problem at all. 

What do you think?

**Comment by bamorim** [comment]
Guys, sorry for late response and thank you so much for the answers and counter proposals.

@johantell okay, the argument about the problems when you forget to define the repo is fair enough. But I'm not sure using false is better.

I'll try to address the problems you pointed in the end

## My view on Ecto

My way of thinking about Ecto is that Repo is the one that depends on the Schema. That is, Schema is the "core" of Ecto and Repo is an "added" feature on top of that that defines an interface for fetching schemas.

That being said, Ecto kind of let you start with just a schema and then, sometime, **opt-in** for a Repo. For example, one could create a lib that defines a bunch of `Ecto.Schema` but doesn't implement any Repo. That could be useful for reusing some concepts.

(Ok, one flaw of my point of view is that Ecto schema contains the table name, but I think that is just a legacy left-over that they couldn't change without breaking a lot of things then moving to Ecto 3, but I'm just guessing here)

With that being said, it seems strange that ExMachina requires a Repo by default and then let's you "opt-out" by passing `repo: false`. It seems counter intuitive for me.

## On `ExMachina.EctoSchema` proposal

Also, @germsvel, although your solution is simple and easy to deploy, my concern is that it will introduce another concept (a new module) to be used when you don't have a Repo.

The problem is that the path of moving from no-repo to with-repo now requires you to change the module you use from `ExMachina.EctoSchema` to `ExMachina.Ecto` (which the naming doesn't help a lot, as you said) and then add the `repo: MyRepo` to the keyword list after that.

It also makes it kind of weird in terms of documentation and usability.

## My new proposal

Okay, if I understood you guys correctly, the problem was that if you forget to add `repo: X` to something, then you can run into some strange problems. The only problem I can think of is trying to call `insert/2` or `insert_list/3` and receiving a weird error like `(UndefinedFunctionError) function MyFactory.insert/2 is undefined or private`, right?

What if instead of not defining the `insert/2` and `insert_list/3` we actually define it but make it raise when called? Something like:

```
You tried to call insert on a ExMachina.Ecto factory but you haven't defined `:repo` on that.
Check link-to-documentation to know more about it
```

Do you think this solves your problems with just not setting `:repo`?

**Comment by johantell** [comment]
@bamorim Absolutely. We might even do a validation check compile time to make sure the option is set to something and raise with a similar helpful message :)

**Comment by col** [comment]
I ran into this issue this week and decided to put a PR together. Happy to change the implementation to meet the consensus but my preferred approach is the same as @bamorim's proposal.

**Comment by bamorim** [comment]
Thanks @col for taking the time! I was a little bit busy and away from Elixir for a while.

**Comment by germsvel** [comment]
Closing this since it was resolved by #370 and is available in [version 2.4.0](https://github.com/thoughtbot/ex_machina/blob/master/CHANGELOG.md#240)
